### DISCUSSION TITLE
Handling `Fn` in `DeletionPolicy`/`UpdateReplacePolicy`

### DISCUSSION THREAD

**Discussion by endgame** [jcouyang/dhall-aws-cloudformation#16] [Ideas]
`DeletionPolicy` is currently implemented as an enum:
```dhall
< Delete | Retain | Snapshot >
```
In some of our stacks, we find ourselves wanting to set different policies on resources depending on conditional expressions involving `Fn::If`.

We currently work around this by declaring the resource and then using ` ⫽ { UpdateReplacePolicy = whatever }` to override the record field, but a native option might be nicer. I'm thinking we may want to extend this enum to something like

```dhall
< Delete | Retain | Snapshot | CfnText CfnText >
```

What do you think?

**Comment by jcouyang** [comment]
i'm not sure if `fn` actually work in `DeletionPolicy`? doc only mentioned [ these values](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-attribute-deletionpolicy.html)...(interesting, didn't notice when there is one more `RetainExceptOnCreate`)

my preference is to avoid using cf's intrinsic function unless there is no choice, because those fn resolve at runtime, which defeated the purpose of converting to dhall, where can type check config as much as possible

 I'm not really against extending it just want to see if there is really a valid use case...so is it possible to convert `fn::if` to dhall expr, unless the value you want to merge is not available at compile time?

**Reply by endgame** [reply]
It is permissible, but you have to add the [`AWS::LanguageExtensions` transform](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/function-refs-in-policy-attributes.html).

We want to use this to make resources that self-delete when the template is instantiated for a staging or test environment, but retain when instantiated to a production environment.

**Reply by jcouyang** [reply]
for different env what i would do is `merge {dev= DeletionPolicy.Delete, prod = DeletionPolicy.Retain} env:current_env`
then pass env when compile the dhall file `env current_env=<dev|prod>.dev dhall-to-yaml <<< ./template.dhall`

but anyway this is just my preference, if you still prefer aws fn::if welcome to send a PR

**Reply by endgame** [reply]
That makes sense, but unfortunately all of the tooling we've built assumes a single template which accepts the stage as a parameter. Perhaps it should not have been this way, but it is. I might look at sending a PR; we are currently making do with record updates.
